### DISCUSSION TITLE
Adding Windows Nodes to cluster

### DISCUSSION THREAD

**Discussion by chzbrgr71** [prometheus-operator/kube-prometheus#624] [General] [answered]
I'm working on getting Windows K8s integrated with my kube-prom deployment. I've deployed a pretty standard AKS cluster and kube-prometheus deployment. 

I've added a Windows node pool and used [windows_exporter](https://github.com/prometheus-community/windows_exporter) to get a node exporter running on Windows. I have it working and I now see metrics on each of the Windows nodes at :9100/metrics. I can give more detail on how I got this working in k8s if needed. 

Now I just need to get my prometheus to scrape metrics from these nodes. Is there any special config needed so that these endpoints are also included? When I add Linux nodepools, they are automatically picked up, but these Windows nodes are not.

**Comment by brancz** [comment, marked as answer]
You will need to add a `Service` and `ServiceMonitor` to discover and scrape the targets.

**Comment by k-modi** [comment]
Hi, I'm also trying to do the same. Given that the windows exporter is running directly on the windows node (not in a container), how did you manage to expose those endpoints via service and service-monitor? Can you please share? thank you so much!

**Comment by k-modi** [comment]
Hi @chzbrgr71 I'm curious to find out how you managed to get the windows-exporter running in K8s.

My approach:
- Created a docker image running windows-exporter,
- Created a daemon-set to run this image on all windows nodes in my K8s cluster
- Created a service for all of pods resulting from the daemon-set, and
- Finally, created a service-monitor to work with the service.

are you using a different approach?

**Comment by chzbrgr71** [comment]
I created a PR with a similar approach, but it does not work as expected. You can read the full discussion below, but the container only gets metrics from the OS in the container, not the host OS. https://github.com/prometheus-community/windows_exporter/pull/581

**Reply by andrewpurdin** [reply]
@chzbrgr71 @k-modi sorry to resurrect an old thread, but did either of you get anything working just to scrape windows nodes? It seems to me that there is no easy way to dynamically add windows hosts as scrape targets for prometheus as they come online, and it also seems that the daemonset approach will require privileged containers before it will function properly.

I was wondering if you could use a webhook that would trigger when a windows node joined the cluster to update the prometheus scrape config with the IP to hit for metrics, but I've never tried anything like that before.

**Comment by paulfantom** [comment]
We have a WIP addon exactly for this in https://github.com/prometheus-operator/kube-prometheus/pull/911

**Reply by andrewpurdin** [reply]
Thanks @paulfantom! That addon definitely looks like a step in the right direction, however we are using autoscaling, so we won't be able to hardcode target IPs in the array.
